### PR TITLE
[Android] Fix Border RoundRectangle CornerRadius issue

### LINKED ISSUE #14276: RoundRectangle StrokeShape produces a different result on iOS vs Android
### Description

https://github.com/dotnet/maui/pull/10964 fixed a number of issues but it looks like there are still a few small areas that need to be fixed

The following `XAML`

```XAML
 <VerticalStackLayout>
        <Border StrokeThickness="20"
                            Stroke="Silver"
                            Background="White"
                            Margin="0"
                            Padding="0">
            <Border.StrokeShape>
                <RoundRectangle CornerRadius="30" />
            </Border.StrokeShape>
            <Grid HeightRequest="30" BackgroundColor="Green">
                <StackLayout Orientation="Horizontal">
                    <Border Padding="0" 
                                       BackgroundColor="Maroon"
                                       HeightRequest="40"
                                       WidthRequest="40" VerticalOptions="Center" >
                        
                        <Border.StrokeShape>
                            <RoundRectangle CornerRadius="20" />
                        </Border.StrokeShape>
                    </Border>
                    <Label Text="Parameters"
                                       VerticalOptions="FillAndExpand" HorizontalOptions="FillAndExpand"
                                       FontSize="Medium" FontFamily="Bold"
                                       HorizontalTextAlignment="Center" VerticalTextAlignment="Center" />
                    <Border Padding="0" Margin="0,0,-20,0"
                                       BackgroundColor="Blue"
                                       HeightRequest="40"
                                       WidthRequest="40" VerticalOptions="Center" >
                    </Border>
                </StackLayout>
            </Grid>
        </Border>
    </VerticalStackLayout>
```

Produces

image

I only tested this on iOS and Android.

### LINKED ISSUE #14936: [regression/8.0-preview 4] Border with RoundRectangle Corner Radius values are smaller than they were
### Description

In .NET 8 P4, on Android specifically, the RoundRectangle stroke shape with a corner radius of '10' is too small.  One thought is that maybe the pixel / dpi scaling calculation has been lost on android?

Android (now, on .NET 8 P4):
image

Windows looks fine still (and android used to look like this too):
image

### Steps to Reproduce

`<Border StrokeShape='{RoundRectangle CornerRadius=10}' Background="Blue" WidthRequest="200" HeightRequest="200" />`

### Link to public reproduction project repository

N/A

### Version with bug

8.0 previews

### Last version that worked well

7.0 (current)

### Affected platforms

Android

### Affected platform versions

All

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Fix **Border** RoundRectangle CornerRadius issue on Android.

Fixes #14936, same corner radius in all the platforms. In the following capture, the top Border is on Windows and the bottom one on Android.
fix-corner

Fixes #14276, on the right side the Android version with the issue:
image
With the changes:
fix-14276

### Issues Fixed

Fixes #14936
Fixes #14276

### CODE REVIEW

File: src/Controls/src/Core/HandlerImpl/RoundRectangle.Impl.cs [hunk 1]
```diff
@@ -51,10 +52,14 @@ public override PathF GetPath()
 			float w = (float)(width - StrokeThickness);
 			float h = (float)(height - StrokeThickness);
 
-			float topLeftCornerRadius = (float)CornerRadius.TopLeft;
-			float topRightCornerRadius = (float)CornerRadius.TopRight;
-			float bottomLeftCornerRadius = (float)CornerRadius.BottomLeft;
-			float bottomRightCornerRadius = (float)CornerRadius.BottomRight;
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/HandlerImpl/RoundRectangle.Impl.cs:57]: Reviewing why we need this changes and if have more sense to apply it at other level. We have not used density before.

Comment (jsuarezruiz) [reply src/Controls/src/Core/HandlerImpl/RoundRectangle.Impl.cs:57]: Maybe related to https://github.com/dotnet/maui/commit/982d809eb9567f75ea810a3d7aedb6aac8f8284c#diff-9962b2ab4d4a0eb4922307e668bfe7d868c4fac7919a9463a3cba1859a6de86fL417-L423

File: src/Controls/src/Core/Shapes/RoundRectangle.cs [hunk 2]
```diff
@@ -66,10 +67,14 @@ public override PathF GetPath()
 			float w = (float)(width - StrokeThickness);
 			float h = (float)(height - StrokeThickness);
 
-			float topLeftCornerRadius = (float)CornerRadius.TopLeft;
-			float topRightCornerRadius = (float)CornerRadius.TopRight;
-			float bottomLeftCornerRadius = (float)CornerRadius.BottomLeft;
-			float bottomRightCornerRadius = (float)CornerRadius.BottomRight;
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
Comment (Redth) [inline src/Controls/src/Core/Shapes/RoundRectangle.cs:72]:
```suggestion
			density = Handler?.MauiContext?.Context?.GetDisplayDensity()
	?? (float)DeviceDisplay.MainDisplayInfo.Density;
```

File: src/Controls/src/Core/Shapes/RoundRectangle.cs [hunk 3]
```diff
@@ -101,10 +106,14 @@ internal PathF GetInnerPath(float strokeWidth)
 			float w = (float)(width - StrokeThickness);
 			float h = (float)(height - StrokeThickness);
 
-			float topLeftCornerRadius = (float)Math.Max(0, CornerRadius.TopLeft - strokeWidth);
-			float topRightCornerRadius = (float)Math.Max(0, CornerRadius.TopRight - strokeWidth);
-			float bottomLeftCornerRadius = (float)Math.Max(0, CornerRadius.BottomLeft - strokeWidth);
-			float bottomRightCornerRadius = (float)Math.Max(0, CornerRadius.BottomRight - strokeWidth);
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
Comment (Redth) [inline src/Controls/src/Core/Shapes/RoundRectangle.cs:111]:
```suggestion
			density = Handler?.MauiContext?.Context?.GetDisplayDensity()
	?? (float)DeviceDisplay.MainDisplayInfo.Density;
```

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Windows.cs [hunk 4]
```diff
@@ -394,6 +394,18 @@ public static async Task<CanvasBitmap> AssertColorAtPointAsync(this FrameworkEle
 			return bitmap.AssertColorAtPoint(expectedColor, x, y);
 		}
 
+		public static async Task<CanvasBitmap> AssertColorsAtPointsAsync(this FrameworkElement view, Graphics.Color[] colors, Graphics.Point[] points, IMauiContext mauiContext)
+		{
+			var bitmap = await view.ToBitmap(mauiContext);
```
[the remark this answers is not in the document]
Comment (jstedfast) [reply src/TestUtils/src/DeviceTests/AssertionExtensions.Windows.cs:399]: yes, it is intentionally like all of the other helper methods that do the same.

**Comment by jstedfast** [comment]
At this point I just need to adjust the Points in the DeviceTests to work for iOS.

**Comment by jstedfast** [comment]
Looks like the iOS Border tests are passing now, but oddly it's getting other errors.

**Comment by jstedfast** [comment]
Seems like random tests are failing

**Review by jstedfast — DISMISSED** [review]

**Review by PureWeen — DISMISSED** [review]
Not sure if this is related to this PR but I noticed in android the border sizing doesn't seem correct. 

I merged in your iOS border samples and notice that android has an odd black border between the red and blue. 

image

**Review by jstedfast — APPROVED** [review]

**Comment by jstedfast** [comment]
I'm beginning to wonder if the density calculation is wrong. It seems like MauiDrawable.Android.cs should pass in `_strokeThickness / _density` instead of `_strokeThickness` because `strokeThickness` has already been multiplied by `_density` but all of the path calculations assume "Maui coordinates" instead of "Android coordinates".

**Comment by jstedfast** [comment]
Replace by https://github.com/dotnet/maui/pull/17087

**Pull request closed without merging**